### PR TITLE
Terraform Support for Dataproc Batches.

### PR DESCRIPTION
**Pull request by animeshnandanwar**
**Release Note Template for Downstream PRs (will be copied)**

```release-note:REPLACEME

```

### CODE REVIEW

File: mmv1/products/dataproc/Batch.yaml [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+  status: !ruby/object:Api::OpAsync::Status
+    path: 'done'
+    complete: true
+    allowed:
+      - true
+      - false
+  error: !ruby/object:Api::OpAsync::Error
+    path: 'error'
+    message: 'message'
+autogen_async: false
```
Comment (melinath) [inline mmv1/products/dataproc/Batch.yaml:45]: Currently the tests are failing because there is no Operation handling code for this resource - is there a particular reason that this can't be autogenerated, or is this just copied from somewhere?

File: mmv1/products/dataproc/Batch.yaml [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+    path: 'response'
+  status: !ruby/object:Api::OpAsync::Status
+    path: 'done'
+    complete: true
+    allowed:
+      - true
+      - false
+  error: !ruby/object:Api::OpAsync::Error
+    path: 'error'
+    message: 'message'
```
Comment (melinath) [inline mmv1/products/dataproc/Batch.yaml:44]: Most of these fields are unused - see https://googlecloudplatform.github.io/magic-modules/develop/resource-reference/#async for more details.

```suggestion

timeouts: !ruby/object:Api::Timeouts
  insert_minutes: 10
  update_minutes: 5
  delete_minutes: 5
async: !ruby/object:Api::OpAsync
  operation: !ruby/object:Api::OpAsync::Operation
    base_url: '{{op_id}}'
```

**Comment by melinath** [comment]
I'm marking this as ready for review so it will show up in my review queue.

File: mmv1/products/dataproc/Batch.yaml [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+    url_param_only: true
+    immutable: true
+    description: |
+      The location in which the batch will be created in.
+  - !ruby/object:Api::Type::String
+    name: 'batchId'
+    url_param_only: true
+    immutable: true
+    description: |
+      The batch Id of the batch.
```
Comment (melinath) [inline mmv1/products/dataproc/Batch.yaml:110]:
```suggestion
      The ID to use for the batch, which will become the final component of the batch's resource name.
      This value must be 4-63 characters. Valid characters are /[a-z][0-9]-/.
```

File: mmv1/products/dataproc/Batch.yaml [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+        output: true
+      - !ruby/object:Api::Type::KeyValuePairs
+        name: 'endpoints'
+        description: |
+          Map of remote access endpoints (such as web interfaces and APIs) to their URIs.
+        output: true
+      - !ruby/object:Api::Type::NestedObject
+        name: 'approximateUsage'
+        description: |
+          Approximate workload resource usage, calculated when the workload completes.
```
Comment (melinath) [inline mmv1/products/dataproc/Batch.yaml:150]: This could be useful to include. There's also a note in the [API docs](https://cloud.google.com/dataproc-serverless/docs/reference/rest/v1/RuntimeInfo) about the meaning of this changing in the future; not sure if you'd want to include that here as well?

I don't think duplicating the pricing link on the subfields is necessary, personally.

```suggestion
          Approximate workload resource usage, calculated when the workload completes (see [Dataproc Serverless pricing](https://cloud.google.com/dataproc-serverless/pricing)).
```

File: mmv1/products/dataproc/Batch.yaml [hunk 5]
```diff
@@ -0,0 +1,504 @@
...
+        output: true
+        properties:
```
Comment (melinath) [inline mmv1/products/dataproc/Batch.yaml:152]: what about milliAcceleratorSeconds and acceleratorType? (here & in currentUsage)

File: mmv1/products/dataproc/Batch.yaml [hunk 6]
```diff
@@ -0,0 +1,504 @@
...
+            output: true
+          - !ruby/object:Api::Type::String
+            name: 'shuffleStorageGbSeconds'
+            description: |
+              Shuffle storage usage in (GB x seconds)
+            output: true
+      - !ruby/object:Api::Type::NestedObject
+        name: 'currentUsage'
+        description: |
+          Approximate workload resource usage, calculated when the workload completes.
```
Comment (melinath) [inline mmv1/products/dataproc/Batch.yaml:166]:
```suggestion
          The usage snapshot represents the resources consumed by a workload at a specified time.
```

File: mmv1/products/dataproc/Batch.yaml [hunk 7]
```diff
@@ -0,0 +1,504 @@
...
+      The state of the batch.
+    output: true
+    values:
+      - :STATE_UNSPECIFIED
+      - :PENDING
+      - :RUNNING
+      - :CANCELLING
+      - :CANCELLED
+      - :SUCCEEDED
+      - :FAILED
```
Comment (melinath) [inline mmv1/products/dataproc/Batch.yaml:206]: Output-only enum fields aren't really meaningful, because the `values` field is unused. I'd recommend something like: 

```suggestion
  - !ruby/object:Api::Type::String
    name: 'state'
    description: |
      The state of the batch. For possible values, see the [API documentation](https://cloud.google.com/dataproc-serverless/docs/reference/rest/v1/projects.locations.batches#State).
    output: true
```

File: mmv1/products/dataproc/Batch.yaml [hunk 8]
```diff
@@ -0,0 +1,504 @@
...
+    name: 'runtimeConfig'
+    description: |
+      Runtime configuration for the batch execution.
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'version'
+        description: |
+          Version of the batch runtime.
+        default_from_api: true
+        diff_suppress_func: cloudDataprocBatchFieldsDiffSuppress
```
Comment (melinath) [inline mmv1/products/dataproc/Batch.yaml:236]: Could be good to have a name that specifies what field this is for, or else what the function checks (SubminorVersionDiffSuppress).

```suggestion
        diff_suppress_func: cloudDataprocBatchRuntimeConfigVersionDiffSuppress
```

File: mmv1/products/dataproc/Batch.yaml [hunk 9]
```diff
@@ -0,0 +1,504 @@
...
+          description: |
+            The state of the batch at this point in history.
+          values:
+            - :STATE_UNSPECIFIED
+            - :PENDING
+            - :RUNNING
+            - :CANCELLING
+            - :CANCELLED
+            - :SUCCEEDED
+            - :FAILED
```
Comment (melinath) [inline mmv1/products/dataproc/Batch.yaml:358]:
```suggestion
        - !ruby/object:Api::Type::String
          name: 'state'
          output: true
          description: |
            The state of the batch at this point in history. For possible values, see the [API documentation](https://cloud.google.com/dataproc-serverless/docs/reference/rest/v1/projects.locations.batches#State).
```

File: mmv1/products/dataproc/Batch.yaml [hunk 10]
```diff
@@ -0,0 +1,504 @@
...
+        - !ruby/object:Api::Type::String
+          name: 'stateMessage'
```
Comment (melinath) [inline mmv1/products/dataproc/Batch.yaml:360]: I think this is not strictly necessary but would be nice for clarity.

```suggestion
          name: 'stateMessage'
          output: true
```

File: mmv1/templates/terraform/constants/cloud_dataproc_batch.go.erb [hunk 11]
```diff
@@ -0,0 +1,15 @@
+/*
+ * Dataproc Batch api apends subminor version to the provided
```
Comment (melinath) [inline mmv1/templates/terraform/constants/cloud_dataproc_batch.go.erb:2]:
```suggestion
 * Dataproc Batch api appends subminor version to the provided
```

File: mmv1/templates/terraform/constants/cloud_dataproc_batch.go.erb [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+ * version. We are suppressing this server generated subminor.
+ */
+func cloudDataprocBatchFieldsDiffSuppressFunc(old, new string) bool {
```
Comment (melinath) [inline mmv1/templates/terraform/constants/cloud_dataproc_batch.go.erb:5]: Please add some [unit tests](https://googlecloudplatform.github.io/magic-modules/develop/test/test/#add-unit-tests) for this function to make sure it's doing what you expect in all cases

File: mmv1/templates/terraform/decoders/cloud_dataproc_batch.go.erb [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+               propertiesCopy := make(map[string]interface{})
+               for k, v := range properties {
+                  propertiesCopy[k] = v
+               }
+               rconfig["effectiveProperties"] = propertiesCopy
+
+               // Update properties back to original client set properties
+               originalPropertiesCopy := make(map[string]interface{})
+               originalProperties := d.Get("runtime_config.0.properties").(interface{}).(map[string]interface{})
+               for k, v := range originalProperties {
```
Comment (melinath) [inline mmv1/templates/terraform/decoders/cloud_dataproc_batch.go.erb:30]: I'm not sure the copying here is strictly necessary but it shouldn't hurt anything even if it's not. 👍

File: mmv1/third_party/terraform/acctest/bootstrap_test_utils.go [hunk 14]
```diff
@@ -910,7 +911,25 @@ func BootstrapSharedCaPoolInLocation(t *testing.T, location string) string {
...
+}
+
+func BootstrapSubnetWithFirewallForDataprocBatches(t *testing.T, testId string, subnetName string) string {
+	networkName := BootstrapSharedTestNetwork(t, testId)
+	subnetworkName := BootstrapSubnetForDataprocBatches(t, subnetName, networkName)
+	BootstrapFirewallForDataprocSharedNetwork(t, subnetName, networkName)
+	return subnetworkName
+}
+
+func BootstrapCustomSubnet(t *testing.T, subnetName string, networkName string, subnetOptions map[string]interface{}) string {
```
Comment (melinath) [inline mmv1/third_party/terraform/acctest/bootstrap_test_utils.go:932]: maybe `BootstrapSubnetWithOptions` or `BootstrapSubnetWithOverrides`?

File: mmv1/third_party/terraform/acctest/bootstrap_test_utils.go [hunk 15]
```diff
@@ -932,13 +951,19 @@ func BootstrapSubnet(t *testing.T, subnetName string, networkName string) string
...
 
-		subnetObj := map[string]interface{}{
+		defaultSubnetObj := map[string]interface{}{
 			"name":        subnetName,
 			"region ":     region,
 			"network":     networkUrl,
 			"ipCidrRange": "10.77.0.0/20",
 		}
 
+		subnetObj := defaultSubnetObj
```
Comment (melinath) [inline mmv1/third_party/terraform/acctest/bootstrap_test_utils.go:961]: Why is this necessary?

**Review by melinath — CHANGES_REQUESTED** [review]
The tests are failing in beta because they get a 404 from the create endpoint. For example:

```
POST /v1beta2/projects/analog-ace-309318/locations/us-central1/batches?alt=json&batchId=tf-test-batch28n9wsuk9x
```

They're not failing when I run them locally in GA, though. (Still waiting to see whether they pass.) Possibly this resource doesn't exist in the beta API?

We would need this resource to be supported in the beta API, or for the beta API to be removed from provider.yaml (if that wouldn't have negative impacts on AutoscalingPolicy - it looks like there aren't any beta-only features on that resource so it _should_ be fine?)

It would be great to also address the missing tests identified by the CI pipeline.

File: mmv1/templates/terraform/decoders/cloud_dataproc_batch.go.erb (hunk 13 above)
Comment (animeshnandanwar) [reply mmv1/templates/terraform/decoders/cloud_dataproc_batch.go.erb:30]: Ack

File: mmv1/templates/terraform/constants/cloud_dataproc_batch.go.erb [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+func CloudDataprocBatchRuntimeConfigVersionDiffSuppressFunc(old, new string) bool {
+        if old == "" || strings.HasPrefix(new, old) {
+          return true
+        }
+
+        return false
+}
+
+func CloudDataprocBatchRuntimeConfigVersionDiffSuppress(_, old, new string, d *schema.ResourceData) bool {
+        return cloudDataprocBatchFieldsDiffSuppressFunc(old, new)
```
Comment (melinath) [inline mmv1/templates/terraform/constants/cloud_dataproc_batch.go.erb:14]:
```suggestion
        return CloudDataprocBatchRuntimeConfigVersionDiffSuppressFunc(old, new)
```

File: mmv1/products/dataproc/Batch.yaml [hunk 17]
```diff
@@ -0,0 +1,511 @@
...
+    name: 'runtimeConfig'
+    description: |
+      Runtime configuration for the batch execution.
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'version'
+        description: |
+          Version of the batch runtime.
+        default_from_api: true
+        diff_suppress_func: cloudDataprocBatchRuntimeConfigVersionDiffSuppress
```
Comment (melinath) [inline mmv1/products/dataproc/Batch.yaml:249]:
```suggestion
        diff_suppress_func: CloudDataprocBatchRuntimeConfigVersionDiffSuppress
```

**Review by melinath — CHANGES_REQUESTED** [review]

File: mmv1/third_party/terraform/services/dataproc/resource_dataproc_batch_test.go.erb [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+   cases := map[string]struct {
+      Old, New           string
+      ExpectDiffSuppress bool
+   }{
+      "empty_default_version": {
+         Old:                "",
+         New:                "2.2.100",
+         ExpectDiffSuppress: true,
+      },
+      // Additional cases excluded for brevity
```
Comment (melinath) [inline mmv1/third_party/terraform/services/dataproc/resource_dataproc_batch_test.go.erb:19]: Please add additional cases

**Review by melinath — CHANGES_REQUESTED** [review]
I made a few minor tweaks to clean up the previous compilation errors - hope you don't mind! I'm waiting for the acceptance tests to run, but in the meantime some additional test cases would be great.

**Review by melinath — CHANGES_REQUESTED** [review]
A couple additional things from the acceptance test run! The tests are failing in beta because they get a 404 from the create endpoint. For example:

POST /v1beta2/projects/analog-ace-309318/locations/us-central1/batches?alt=json&batchId=tf-test-batch28n9wsuk9x
They're passing when I run them locally in GA, though. Possibly this resource doesn't exist in the beta API?

We would need this resource to be supported in the beta API, or for the beta API to be removed from provider.yaml (if that wouldn't have negative impacts on AutoscalingPolicy - it looks like there aren't any beta-only features on that resource so it should be fine?)

It would be great to also address the missing tests identified by the CI pipeline.

**Comment by zli82016** [comment]
> A couple additional things from the acceptance test run! The tests are failing in beta because they get a 404 from the create endpoint. For example:
> 
> ```
> POST /v1beta2/projects/analog-ace-309318/locations/us-central1/batches?alt=json&batchId=tf-test-batch28n9wsuk9x
> ```
> 
> They're passing when I run them locally in GA, though. Possibly this resource doesn't exist in the beta API?
> 
> We would need this resource to be supported in the beta API, or for the beta API to be removed from provider.yaml (if that wouldn't have negative impacts on AutoscalingPolicy - it looks like there aren't any beta-only features on that resource so it should be fine?)
> 

@melinath , is it a breaking change to remove the beta API for AutoscalingPolicy resource?

**Comment by zli82016** [comment]
I opened a new PR
https://github.com/GoogleCloudPlatform/magic-modules/pull/11750
